internal/mongo/watch_producer.go

```diff
@@ -31,49 +32,95 @@ func (w *WatchProducer) GetProducer(o ...WatchOption) ChangeEventProducer {
 			pipeline = append(customElements, pipeline...)
 		}
 
-		opts := &options.ChangeStreamOptions{
-			BatchSize:            &config.batchSize,
-			MaxAwaitTime:         &config.maxAwaitTime,
-			StartAtOperationTime: config.startAtOperationTime,
-		}
-		if config.fullDocumentEnabled {
-			opts.SetFullDocument(options.UpdateLookup)
-		}
-		if len(config.resumeAfter) != 0 {
-			opts.SetResumeAfter(config.resumeAfter)
-		}
-
-		cursor, err := w.collection.Watch(ctx, pipeline, opts)
+		cursor, err := w.watch(ctx, pipeline, config, nil)
 		if err != nil {
-			w.logger.Error("Mongo client: An error has occured while watching collection", logger.String("collection", w.collection.Name()), logger.Error("error", err))
+			w.logger.Error("Mongo client: An error has occured while trying to watch collection", logger.String("collection", w.collection.Name()), logger.Error("error", err))
 			return nil, err
 		}
 
 		var events = make(chan *ChangeEvent)
 
 		go func() {
-			defer cursor.Close(ctx)
+			ctx, cancel := context.WithCancel(ctx)
+			defer cancel()
 			defer close(events)
-
 			for {
-				w.sendEvents(ctx, cursor, events)
+				resumeToken, err := w.sendEvents(ctx, cursor, events)
+				w.logger.Error("Mongo client : An error has occured while watching collection", logger.String("collection", w.collection.Name()), logger.Error("error", err))
+
+				cursor.Close(ctx)
+
+				w.logger.Info("Mongo client : Retry to watch collection", logger.String("collection", w.collection.Name()), logger.Any("token", resumeToken))
+
+				cursor, err = w.watch(ctx, pipeline, config, resumeToken)
+				if err != nil {
+					w.logger.Error("Mongo client : An error has occured while retrying to watch collection", logger.String("collection", w.collection.Name()), logger.Error("error", err))
+					break
+				}
 			}
 		}()
 
 		return events, nil
 	}
 }
 
-func (w *WatchProducer) sendEvents(ctx context.Context, cursor DriverCursor, events chan *ChangeEvent) {
+func (w *WatchProducer) watch(ctx context.Context, pipeline bson.A, config *WatchConfig, resumeToken bson.Raw) (cursor StreamCursor, err error) {
+	// choose the resume token to use
+	var resumeAfter interface{}
+	if resumeToken != nil {
+		resumeAfter = resumeToken
+	} else if len(config.resumeAfter) > 0 {
+		resumeAfter = config.resumeAfter
+	}
+	// retries loop
+	attempt := int32(0)
+	for {
+		opts := &options.ChangeStreamOptions{
+			BatchSize:            &config.batchSize,
+			MaxAwaitTime:         &config.maxAwaitTime,
+			StartAtOperationTime: config.startAtOperationTime,
+		}
+		if config.fullDocumentEnabled {
+			opts.SetFullDocument(options.UpdateLookup)
+		}
+		if resumeAfter != nil {
+			opts.SetResumeAfter(resumeAfter)
+		}
+
+		cursor, err = w.collection.Watch(ctx, pipeline, opts)
+		if err == nil {
+			break
+		}
+		if attempt >= config.maxRetries {
+			break
```

Comment:
Maybe we could log an informational message here to let people know that the maximum attempt number has been reached